csrc/mmdeploy/backend_ops/tensorrt/modulated_deform_conv_v3/trt_deform_conv_v3_kernel.cu

```diff
@@ -0,0 +1,185 @@
+    const int width_in, const int height_out, const int width_out, const scalar_t offset_scale) {
+  CUDA_KERNEL_LOOP(index, num_kernels) {
+    int _temp = index;
+    const int c_col = _temp % group_channels;
+    _temp /= group_channels;
+    const int sampling_index = _temp;
+    const int g_col = _temp % group;
+    _temp /= group;
+    const int p0_w = ((dilation_w * (kernel_w - 1)) >> 1) - pad_w + (_temp % width_out) * stride_w;
+    _temp /= width_out;
+    const int p0_h = ((dilation_h * (kernel_h - 1)) >> 1) - pad_h + (_temp % height_out) * stride_h;
+    _temp /= height_out;
+    const int b_col = _temp;
+
+    const int input_size = height_in * width_in;
+    scalar_t *data_col_ptr = data_col + index;
+    const int kernel_size = kernel_h * kernel_w;
+    int data_weight_ptr = sampling_index * kernel_size;
+    int data_loc_w_ptr = data_weight_ptr << 1;
+    const int qid_stride = group * group_channels;
+    scalar_t col = 0;
+    const scalar_t *data_im_ptr = data_im + b_col * input_size * qid_stride;
+    // top-left
+    const scalar_t p0_w_ = p0_w - ((dilation_w * (kernel_w - 1)) >> 1) * offset_scale;
+    const scalar_t p0_h_ = p0_h - ((dilation_h * (kernel_h - 1)) >> 1) * offset_scale;
+    for (int i = 0; i < kernel_w; ++i) {
+      for (int j = 0; j < kernel_h; ++j) {
+        const scalar_t offset_w = data_offset[data_loc_w_ptr];
+        const scalar_t offset_h = data_offset[data_loc_w_ptr + 1];
+        const scalar_t loc_w = p0_w_ + (i * dilation_w + offset_w) * offset_scale;
+        const scalar_t loc_h = p0_h_ + (j * dilation_h + offset_h) * offset_scale;
+        const scalar_t weight = data_mask[data_weight_ptr];
+        if (loc_h > -1 && loc_w > -1 && loc_h < height_in && loc_w < width_in) {
+          col += dcnv3_im2col_bilinear(data_im_ptr, height_in, width_in, group, group_channels,
+                                       loc_h, loc_w, g_col, c_col) *
+                 weight;
+        }
+        data_weight_ptr += 1;
+        data_loc_w_ptr += 2;
+      }
+    }
+    *data_col_ptr = col;
+  }
+}
+
+template <typename scalar_t>
+void dcnv3_im2col_cuda(cudaStream_t stream, const scalar_t *data_im, const scalar_t *data_offset,
+                       const scalar_t *data_mask, scalar_t *data_col, const int kernel_h,
+                       const int kernel_w, const int stride_h, const int stride_w, const int pad_h,
+                       const int pad_w, const int dilation_h, const int dilation_w, const int group,
+                       const int group_channels, const int batch_n, const int height_in,
+                       const int width_in, const int height_out, const int width_out,
+                       const scalar_t offset_scale) {
+  const int num_kernels = batch_n * height_out * width_out * group * group_channels;
+  const int num_actual_kernels = batch_n * height_out * width_out * group * group_channels;
+  const int num_threads = CUDA_NUM_THREADS;
+  dcnv3_im2col_gpu_kernel<scalar_t>
```

Comment:
So this dcnv3 plugin does not include convolution?

Reply:
Yes.

Reply:
The behavior is different from the name 'deform_**conv**_v3'. It would be cool if we add some comments in docs and code.

Reply:
That's a good idea. The codes were borrow from [internimage](https://github.com/OpenGVLab/InternImage/blob/4fb17721a0f9ab9fb28c7ed48ac1667a247c6da4/classification/ops_dcnv3/src/cuda/dcnv3_cuda.cu#L71)